Gemfile

```diff
@@ -31,7 +31,7 @@ gem "bootstrap_form", ">= 4.2.0"
 group :development, :test do
   gem "byebug", platforms: [:mri, :mingw, :x64_mingw]
   gem "rspec-rails"
-  gem "capybara", ">= 2.15"
+  gem "capybara"
   gem "factory_bot_rails"
```

Comment:
gem này dùng là  gì?

Reply:
Feature spec sẽ test ứng dụng web bằng cách mô phỏng một người dùng thực sẽ tương tác với trang web của bạn như thế nào.
=> Gem `Capybara` được thiết kế để mô phỏng các request của browser với HTTP. Nó chủ yếu sẽ gửi nội dung HTML.
VD: đoạn code sau sẽ test khi tạo người dung
<RSpec.feature "User management", :type => :feature do
  scenario "Admin creates a new user" do
    visit "/user/new"
 
    fill_in "Name", with: "First author" # Điền tên user vào textfiled Name
    fill_in "Age", with: 30 # Điền tuổi user vào textfield Age
    click_button "Create user" # Click vào button 'Create user'
 
    expect(page).to have_text("User was successfully created.")
  end
end>
*** Một số thuộc tính sử dụng trong Gem này:
- Clicking links and buttons:
<click_link('id-of-link')
click_link('Link Text')
click_button('Save')
click_on('Link Text') # có thể click vào link hoặc button
click_on('Button Value')>

- Tương tác với forms:
<fill_in('Full Name', :with => 'huu phuc')
fill_in('Password', :with => '12345678')
fill_in('Description', :with => 'A handsome boy...')
choose('A Radio Button')
check('A Checkbox')
uncheck('A Checkbox')
attach_file('Image', '/path/to/image.jpg')
select('Option', :from => 'Select Box')>

- Querying:
Capybara có nhiều tùy chọn cho việc truy vấn trang về sự tồn tại của các thành phần nào đó, làm việc và thao tác với các thành phần đó.
<page.has_selector?('table tr')
page.has_selector?(:xpath, '//table/tr')
page.has_xpath?('//table/tr')
page.has_css?('table tr.foo')
page.has_content?('foo')>

- Tìm kiếm các thành phần trong trang web:
<find_field('First Name').value
find_link('Hello', :visible => :all).visible?
find_button('Send').click
find(:xpath, "//table/tr").click
find("#overlay").find("h1").click
all('a').each { |a| a[:href] }>
- Scoping:
Gem này cho phép ta truy giới hạn những hành động cụ thể, VD như ở đây có cacs tùy chọn cụ thể để điền các tùy chọn test(form, links, buttons):
<within("li#employee") do
  fill_in 'Name', :with => 'Phuc'
end
within(:xpath, "//li[@id='employee']") do
  fill_in 'Name', :with => 'Phuc'
end>
Ngoài ra còn có một số hàm khác dành cho các fieldset hoặc table, định danh bởi id hoặc đoạn text của fieldset hay table đó:
<within_fieldset('Employee') do
  fill_in 'Name', :with => 'Phuc'
end
within_table('Employee') do
  fill_in 'Name', :with => 'Phuc'
end>
- Làm việc với các cửa sổ(tìm kiếm và chuyển đổi cửa sổ trang web):
<facebook_window = window_opened_by do
  click_button 'Like'
end
within_window facebook_window do
  find('#login_email').set('a@example.com')
  find('#login_password').set('qwerty')
  click_button 'Submit'
end>
- thực thi Javascript trong test:
<page.execute_script("$('body').empty()")
page.execute_script("$('button#next').prop('disabled', false)")`
- Modals: có thể chấp nhận, từ chối hay đáp ứng đối với các alert, confirm và prompt
`accept_alert do
  click_link('Show Alert')
end
dismiss_alert do
  click_link('Show Alert')
end
dismiss_confirm do
  click_link('Show Confirm')
end
accept_prompt(with: 'Linus Torvalds') do
  click_link('Show Prompt About Linux')
end>
- Debugging: 
<save_and_open_page  #chụp được ảnh của trang hiện tại
print page.html #lấy được trạng thái hiện tại của DOM>